src/encode.rs

```diff
@@ -0,0 +1,102 @@
+pub const MAX_U32_LENGTH: usize = 5;
+
+#[derive(Debug)]
+pub struct Encoder<'a> {
+    dst: &'a mut Vec<u8>,
+}
+
+impl<'data> Encoder<'data> {
+    pub fn new(dst: &'data mut Vec<u8>) -> Encoder<'data> {
+        Encoder { dst }
+    }
+
+    pub fn byte(&mut self, byte: u8) {
+        self.dst.push(byte);
+    }
+
+    pub fn bytes(&mut self, bytes: &[u8]) {
+        self.usize(bytes.len());
+        self.raw(bytes);
+    }
+
+    pub fn str(&mut self, data: &str) {
+        self.bytes(data.as_bytes())
+    }
+
+    pub fn usize(&mut self, amt: usize) {
+        assert!(amt <= u32::max_value() as usize);
+        self.u32(amt as u32)
+    }
+
+    pub fn u32(&mut self, mut amt: u32) {
+        while amt >= (1 << 7) {
+            self.byte((amt as u8) & 0x7f | 0x80);
+            amt >>= 7;
+        }
+        self.byte(amt as u8);
+    }
+
+    pub fn i32(&mut self, val: i32) {
+        self.i64(val as i64);
+    }
+
+    pub fn i64(&mut self, mut val: i64) {
+        let mut done = false;
+        while !done {
+            let mut byte = (val as i8) & 0x7f;
+            val >>= 7;
```

Comment:
So the `leb128` crate has an existing writer that we could use here, and (at least the reader part of) that crate has had a good amount of optimization work done. Would be worth investigating at minimum.

Reply:
https://github.com/gimli-rs/leb128

Reply:
I actually initially tried to use that, but I ran into one snag. When this implementation encodes sections it reserves 5 bytes for the section length, encodes the section, and then goes back and fills in the size of the section. That may not be the best idea in general, but the leb128 crate didn't have a way (I think?) to encode an integer to the maximum width, generating unnecessary trailer bytes.

Do you think we should abandon this strategy entirely and just slosh around buffers? (I haven't benchmarked anything yet). Or do you think we should add that functionality to leb128 crate?

Reply:
Ok sure thing, I've filed an issue at https://github.com/gimli-rs/leb128/issues/6 to track it for now, and I'll put a FIXME in the code to stop hardcoding.